jupyter_server_fileid/manager.py

```diff
@@ -37,41 +36,167 @@ def wrapped(self, *args, **kwargs):
     return decorator
 
 
-class AbstractFileIdManager(ABC):
-    @abstractmethod
-    def get_id(self, path: str) -> str:
-        ...
+class AbstractFileIdManager(LoggingConfigurable):
+    """
+    Abstract base class for File ID manager implementations. All File ID
+    managers should inherit from this class.
+    """
+
+    root_dir = Unicode(
+        help=("The root being served by Jupyter server. Must be an absolute path."), config=False
+    )
+
+    db_path = Unicode(
+        default_value=default_db_path,
+        help=(
+            "The path of the DB file used by File ID manager implementations. "
+            "Defaults to `jupyter_data_dir()/file_id_manager.db`."
+        ),
+        config=True,
+    )
+
+    @validate("root_dir", "db_path")
+    def _validate_abspath_traits(self, proposal):
+        if proposal["value"] is None:
+            raise TraitError(f"LocalFileIdManager : {proposal['trait'].name} must not be None")
+        if not os.path.isabs(proposal["value"]):
+            raise TraitError(
+                f"LocalFileIdManager : {proposal['trait'].name} must be an absolute path"
+            )
+        return proposal["value"]
 
-    @abstractmethod
-    def get_path(self, id: str) -> str:
-        ...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def index(self, path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_id(self, path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_path(self, id: Union[int, str]) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def move(self, old_path: str, new_path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def copy(self, from_path: str, to_path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def delete(self, path: str) -> None:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def save(self, path: str) -> None:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_handlers_by_action(self) -> Dict[str, Optional[Callable[[Dict[str, Any]], Any]]]:
+        """Returns a dictionary whose keys are contents manager event actions
+        and whose values are lambdas invoked upon receipt of an event of the
+        same action. The lambda accepts the body of the event as its only
+        argument. To ignore an event action, set the value to `None`."""
```

Comment:
Instead of talking about lambdas, we should use "calllable" ("lambda" is an implementation detail).
```suggestion
        """Returns a dictionary whose keys are contents manager event actions
        and whose values are callables invoked upon receipt of an event of the
        same action. The callable accepts the body of the event as its only
        argument. To ignore an event action, set the value to `None`."""
```